src/Bicep.LangServer/Completions/PublicRegistryModuleMetadataProvider.cs

```diff
@@ -16,69 +17,124 @@ namespace Bicep.LanguageServer.Providers
-        {
-            List<string> moduleNames = new List<string>();
+                    this.cachedModules = modules;
+                    this.lastSuccessfulQuery = DateTime.Now;
+                }
+                finally
+                {
+                    // Throttle unsuccessful requests to avoid spamming the endpoint
+                    await Task.Delay(5000);
 
-            if (!moduleMetadataCache.Any())
-            {
-                await Initialize();
-            }
+                    lock (this.queryingLiveSyncObject)
+                    {
+                        Trace.Assert(this.isQueryingLiveData, "this.isQueryingLiveData should be true");
+                        this.isQueryingLiveData = false;
+                    }
+                }
+            });
+        }
 
-            foreach (var moduleMetadataCacheEntry in moduleMetadataCache)
+        private bool IsCacheExpired()
+        {
+            var expired = this.lastSuccessfulQuery.HasValue && this.lastSuccessfulQuery.Value + this.cacheValidFor < DateTime.Now;
+            if (expired)
             {
-                moduleNames.Add(moduleMetadataCacheEntry.moduleName);
+                Trace.TraceInformation("Public modules cache is expired.");
             }
 
-            return moduleNames;
+            return expired;
         }
 
-        public async Task<IEnumerable<string>> GetVersions(string moduleName)
+        private async Task<ModuleMetadata[]> TryGetModulesLive()
         {
-            List<string> versions = new List<string>();
+            Trace.WriteLine($"Retrieving list of public registry modules...");
 
-            if (!moduleMetadataCache.Any())
+            try
             {
-                await Initialize();
-            }
-
-            ModuleMetadata? metadata = moduleMetadataCache.FirstOrDefault(x => x.moduleName.Equals(moduleName, StringComparison.Ordinal));
+                HttpClient httpClient = new HttpClient();
+                var moduleMetadata = await httpClient.GetStringAsync(LiveDataEndpoint);
+                var metadata = JsonConvert.DeserializeObject<ModuleMetadata[]>(moduleMetadata);
 
-            if (metadata is not null)
+                if (metadata is not null)
+                {
+                    return metadata;
+                }
+                else
+                {
+                    throw new Exception($"List of MCR modules at {LiveDataEndpoint} was empty");
+                }
+            }
+            catch (Exception e)
             {
-                return metadata.tags.OrderDescending();
+                Trace.TraceError(string.Format("Error retrieving MCR modules metadata: {0}", e.Message));
+                return new ModuleMetadata[] { };
             }
+        }
 
-            return Enumerable.Empty<string>();
+        // e.g. "app/dapr-containerapp"
+        public Task<IEnumerable<string>> GetModuleNames()
+        {
+            UpdateCacheAsync();
```

Comment:
[nit] Name this something to indicate it's not always going to refresh the cache - `UpdateCacheIfRequired()`?